pkg/controller/dnsendpoint/nameserver/azure.go

```diff
@@ -0,0 +1,183 @@
+				credsSecret,
+			); err != nil {
+				return nil, errors.Wrap(err, "could not get the creds secret")
+			}
+			azureClient, err := azureclient.NewClientFromSecret(credsSecret)
+			return azureClient, errors.Wrap(err, "error creating Azure client")
+		},
+		resourceGroupName: resourceGroupName,
+	}
+}
+
+type azureQuery struct {
+	getAzureClient    func() (azureclient.Client, error)
+	resourceGroupName string
+}
+
+var _ Query = (*azureQuery)(nil)
+
+// Get implements Query.Get.
+func (q *azureQuery) Get(domain string) (map[string]sets.String, error) {
+	azureClient, err := q.getAzureClient()
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get azure client")
+	}
+	currentNameServers, err := q.queryNameServers(azureClient, domain)
+	return currentNameServers, errors.Wrap(err, "error querying name servers")
+}
+
+// Create implements Query.Create.
+func (q *azureQuery) Create(rootDomain string, domain string, values sets.String) error {
+	azureClient, err := q.getAzureClient()
+	if err != nil {
+		return errors.Wrap(err, "failed to get Azure client")
+	}
+
+	return errors.Wrap(q.createNameServers(azureClient, rootDomain, domain, values), "error creating the name server")
+}
+
+// Delete implements Query.Delete.
+func (q *azureQuery) Delete(rootDomain string, domain string, values sets.String) error {
+	azureClient, err := q.getAzureClient()
+	if err != nil {
+		return errors.Wrap(err, "failed to get Azure client")
+	}
+
+	if len(values) != 0 {
+		// If values were provided for the name servers, attempt to perform a
+		// delete using those values.
+		return errors.Wrap(q.deleteNameServers(azureClient, rootDomain, domain, values), "error deleting the name server")
+	}
+
+	// Since we do not have up-to-date values for the name servers, we need
+	// to query Azure for the current values to use them in the delete.
+	values, err = q.queryNameServer(azureClient, rootDomain, domain)
+	if err != nil {
+		return errors.Wrap(err, "error querying the current values of the name server")
+	}
+	if len(values) == 0 {
+		return nil
+	}
+
+	return errors.Wrap(
+		q.deleteNameServers(azureClient, rootDomain, domain, values),
+		"error deleting the name server with recently read values",
+	)
+}
+
+// deleteNameServers deletes the name servers for the specified domain in the specified managed zone.
+func (q *azureQuery) deleteNameServers(azureClient azureclient.Client, rootDomain string, domain string, values sets.String) error {
+	return azureClient.DeleteRecordSet(context.TODO(), q.resourceGroupName, rootDomain, domain, dns.NS)
+}
+
+// createNameServers creates the name servers for the specified domain in the specified managed zone.
+func (q *azureQuery) createNameServers(azureClient azureclient.Client, rootDomain string, domain string, values sets.String) error {
+	_, err := azureClient.CreateOrUpdateRecordSet(context.TODO(), q.resourceGroupName, rootDomain, domain, dns.NS, q.recordSet(domain, values))
```

Comment:
This does not create the NS records with the correct domain. The `recordSetName` parameter must be relative to the root domain not absolute.

Reply:
Made it relative